hips/tiles/fetch.py

```diff
@@ -0,0 +1,130 @@
+    """Fetch a list of HiPS tiles.
+
+    This function fetches a list of HiPS tiles based
+    on their URLs, which are generated using `hips_survey`
+    and `tile_metas`. The tiles are then fetched asynchronously
+    using urllib or aiohttp.
+
+    Parameters
+    ----------
+    tile_metas : List[HipsTileMeta]
+        Python list of HipsTileMeta
+    hips_survey : `~hips.HipsSurveyProperties`
+        HiPS survey properties
+    progress_bar : bool
+        Show a progress bar for tile fetching and drawing
+    n_parallel : int
+        Number of threads to use for fetching HiPS tiles
+    timeout : int
+        Seconds to timeout for fetching a HiPS tile
+    fetch_package : {'urllib', 'aiohttp'}
+        Package to use for fetching HiPS tiles
+
+    Examples
+    --------
+    >>> from hips import HipsSurveyProperties
+    >>> from hips import fetch_tiles
+    >>> url='http://alasky.unistra.fr/DSS/DSS2Merged/properties',
+    >>> hips_survey = HipsSurveyProperties.fetch(url)
+    >>> tile_indices = [69623, 69627, 69628, 69629, 69630, 69631]
+    >>> tile_metas = []
+    >>> for healpix_pixel_index in tile_indices:
+    ...    tile_meta = HipsTileMeta(
+    ...        order=7,
+    ...        ipix=healpix_pixel_index,
+    ...        frame=hips_survey.astropy_frame,
+    ...        file_format='fits',
+    ...    )
+    ...    tile_metas.append(tile_meta)
+    >>> tiles = fetch_tiles(tile_metas, hips_survey)
+    >>> tiles[0].meta.file_format
+    fits
+    >>> tiles[0].meta.order
+    7
+
+    Returns
+    -------
+    tiles : List[HipsTile]
+        A Python list of HiPS tiles
+    """
+    if fetch_package == 'aiohttp':
+        return tiles_aiohttp(tile_metas, hips_survey, progress_bar)
+    elif fetch_package == 'urllib':
+        return tiles_urllib(tile_metas, hips_survey, progress_bar, n_parallel, timeout)
+    else:
+        raise ValueError(f'Invalid package name: {fetch_package}')
+
+def fetch_tile_urllib(url: str, meta: HipsTileMeta, timeout: float) -> Generator:
+    """Fetch a HiPS tile asynchronously."""
+    with urllib.request.urlopen(url, timeout=timeout) as conn:
+        raw_data = conn.read()
+        return HipsTile(meta, raw_data)
+
+def tiles_urllib(tile_metas: List[HipsTileMeta], hips_survey : HipsSurveyProperties,
+                 progress_bar: bool = False, n_parallel: int = 10,  timeout: int = 10) -> List[HipsTile]:
+    """Generator function to fetch HiPS tiles from a remote URL."""
+    with concurrent.futures.ThreadPoolExecutor(max_workers=n_parallel) as executor:
+        if progress_bar:
+            from tqdm import tqdm
+            tile_metas = tqdm(tile_metas, total=len(tile_metas), desc='Fetching tiles')
+
+        tiles = []
+        for meta in tile_metas:
+            url = hips_survey.tile_url(meta)
+            future = executor.submit(
+                fetch_tile_urllib,
+                url,
+                meta,
+                timeout)
+            tiles.append(future.result())
```

Comment:
I think the problem why for threads it's executed synchronously is here.
You have to have one for loop that submits all the tasks, that will give you a list of futures.

And then only after everything is submitted you can call `.result()` or do some other blocking call that will wait for results to arrive.